### ISSUE TITLE
bizarre behavior with sort and CORE::fake

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14323]
Migrated from [rt.perl.org#123410](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123410) (status was 'resolved')

Searchable as RT123410$

**Comment by p5pRT** [comment]
### From @rjbs

#### Created by @rjbs



&nbsp;  ~$ perl \-MO=Deparse \-E'sort CORE&#8203;::abc\_def\("\*"\)'
&nbsp;  use feature 'current\_sub'\, 'evalbytes'\, 'fc'\, 'say'\, 'state'\, 'switch'\, 'unicode\_strings'\, 'unicode\_eval';
&nbsp;  sort CORE&#8203;::a '\*';
&nbsp;  \-e syntax OK

&nbsp;  ~$ perl \-MO=Deparse \-E'sort CORE&#8203;::abc\_def\_ghi\("\*"\)'
&nbsp;  use feature 'current\_sub'\, 'evalbytes'\, 'fc'\, 'say'\, 'state'\, 'switch'\, 'unicode\_strings'\, 'unicode\_eval';
&nbsp;  sort CORE&#8203;::abc\_d '\*';
&nbsp;  \-e syntax OK

Woah\.

&nbsp;  11&#8203;:51 \<ilmari> it's chopping length\("CORE&#8203;::"\) off the wrong end?
&nbsp;  11&#8203;:51 \<rjbs> ilmari&#8203;: \!\!

This isn't just Deparse\.

&nbsp;  ~$ perl  \-E'use Data&#8203;::Dumper; warn Dumper \[sort CORE&#8203;::abc\_def\("\*"\)\]'
&nbsp;  Undefined sort subroutine "CORE&#8203;::a" called at \-e line 1\.
&nbsp;  ~$ perl  \-E'use Data&#8203;::Dumper; warn Dumper \[sort CORE&#8203;::abc\_def\_ghi\("\*"\)\]'
&nbsp;  Undefined sort subroutine "CORE&#8203;::abc\_d" called at \-e line 1\.

Bug appears in v5\.20\, not in v5\.18\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.21.5:

Configured by rjbs at Mon Oct 20 13:54:04 EDT 2014.

Summary of my perl5 (revision 5 version 21 subversion 5) configuration:
   
  Platform:
    osname=darwin, osvers=14.0.0, archname=darwin-2level
    uname='darwin walrus.local 14.0.0 darwin kernel version 14.0.0: fri sep 19 00:26:44 pdt 2014; root:xnu-2782.1.97~2release_x86_64 x86_64 '
    config_args='-Dprefix=/Users/rjbs/.plenv/versions/21.5 -de -Dusedevel -A'eval:scriptdir=/Users/rjbs/.plenv/versions/21.5/bin''
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-common -DPERL_DARWIN -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -I/opt/local/include',
    optimize='-O3',
    cppflags='-fno-common -DPERL_DARWIN -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -I/opt/local/include'
    ccversion='', gccversion='4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.51)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16, longdblkind=3
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags =' -fstack-protector -L/usr/local/lib -L/opt/local/lib'
    libpth=/usr/local/lib /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/6.0/lib /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib /usr/lib /opt/local/lib
    libs=-lgdbm -ldbm -ldl -lm -lutil -lc
    perllibs=-ldl -lm -lutil -lc
    libc=, so=dylib, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup -L/usr/local/lib -L/opt/local/lib -fstack-protector'



@INC for perl 5.21.5:
    /Users/rjbs/.plenv/versions/21.5/lib/perl5/site_perl/5.21.5/darwin-2level
    /Users/rjbs/.plenv/versions/21.5/lib/perl5/site_perl/5.21.5
    /Users/rjbs/.plenv/versions/21.5/lib/perl5/5.21.5/darwin-2level
    /Users/rjbs/.plenv/versions/21.5/lib/perl5/5.21.5
    .


Environment for perl 5.21.5:
    DYLD_LIBRARY_PATH (unset)
    HOME=/Users/rjbs
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/Users/rjbs/.plenv/versions/21.5/bin:/Users/rjbs/.plenv/libexec:/Users/rjbs/.plenv/plugins/perl-build/bin:/Users/rjbs/bin:/Users/rjbs/.rbenv/shims:/Users/rjbs/.rbenv/bin:/Users/rjbs/.plenv/shims:/Users/rjbs/.plenv/bin:/opt/local/bin:/opt/local/sbin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/opt/X11/bin:/Users/rjbs/code/hla
    PERL_AUTOINSTALL=--skipdeps
    PERL_BADLANG (unset)
    PERL_MAILERS=sendmail:/Users/rjbs/bin/sendmail
    SHELL=/opt/local/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @wolfsage

On Thu\, Dec 11\, 2014 at 11&#8203;:55 AM\, Ricardo SIGNES
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>   ~$ perl \-MO=Deparse \-E'sort CORE&#8203;::abc\_def\("\*"\)'
>   use feature 'current\_sub'\, 'evalbytes'\, 'fc'\, 'say'\, 'state'\, 'switch'\, 'unicode\_strings'\, 'unicode\_eval';
>   sort CORE&#8203;::a '\*';
>   \-e syntax OK
>
>   ~$ perl \-MO=Deparse \-E'sort CORE&#8203;::abc\_def\_ghi\("\*"\)'
>   use feature 'current\_sub'\, 'evalbytes'\, 'fc'\, 'say'\, 'state'\, 'switch'\, 'unicode\_strings'\, 'unicode\_eval';
>   sort CORE&#8203;::abc\_d '\*';
>   \-e syntax OK

Looks like this broke between 5\.19\.0 and 5\.19\.1&#8203;:

&nbsp;   /home/mhorsfall/dpppperls/default/perl\-5\.19\.0/bin/perl5\.19\.0
\-MO=Deparse '\-E\[sort CORE&#8203;::abc\_def\("\*"\)\]' 2>&1
&nbsp;  \-e syntax OK
&nbsp;  \[\(sort CORE&#8203;::abc\_def '\*'\)\];
&nbsp;   /home/mhorsfall/dpppperls/default/perl\-5\.19\.1/bin/perl5\.19\.1
\-MO=Deparse '\-E\[sort CORE&#8203;::abc\_def\("\*"\)\]' 2>&1
&nbsp;  \-e syntax OK
&nbsp;  \[\(sort CORE&#8203;::a '\*'\)\];

&nbsp;  /home/mhorsfall/dpppperls/default/perl\-5\.19\.0/bin/perl5\.19\.0 '\-Euse
File&#8203;::Glob bsd\_glob; use Data&#8203;::Dumper; warn Dumper \[sort
CORE&#8203;::bsd\_glob\("\*"\)\]' 2>&1
&nbsp;  Undefined sort subroutine "CORE&#8203;::bsd\_glob" called at \-e line 1\.
&nbsp;  child exited with value 255
&nbsp;  /home/mhorsfall/dpppperls/default/perl\-5\.19\.1/bin/perl5\.19\.1 '\-Euse
File&#8203;::Glob bsd\_glob; use Data&#8203;::Dumper; warn Dumper \[sort
CORE&#8203;::bsd\_glob\("\*"\)\]' 2>&1
&nbsp;  Undefined sort subroutine "CORE&#8203;::bs" called at \-e line 1\.
&nbsp;  child exited with value 255

\-\- Matthew Horsfall \(alh\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rjbs

\* "Matthew Horsfall \(alh\)" \<wolfsage@&#8203;gmail\.com> \[2014\-12\-11T12&#8203;:00&#8203;:01\]
> Looks like this broke between 5\.19\.0 and 5\.19\.1&#8203;:

Unsurprisingly\, it bisects to 01b5ef509f2ebf466fd7de2c1e7406717bb14332

&nbsp;  commit 01b5ef509f2ebf466fd7de2c1e7406717bb14332
&nbsp;  Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
&nbsp;  Date&#8203;:   Fri Jun 7 20&#8203;:16&#8203;:23 2013 \-0700

&nbsp;      \[perl \#24482\] Fix sort and require to treat CORE&#8203;:: as keyword

Looks like the code assumes that CORE&#8203;::\.\.\. is never bogus?

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Dec 11 10&#8203;:45&#8203;:12 2014\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> \* "Matthew Horsfall \(alh\)" \<wolfsage@&#8203;gmail\.com> \[2014\-12\-11T12&#8203;:00&#8203;:01\]
> > Looks like this broke between 5\.19\.0 and 5\.19\.1&#8203;:
> 
> Unsurprisingly\, it bisects to 01b5ef509f2ebf466fd7de2c1e7406717bb14332
> 
>   commit 01b5ef509f2ebf466fd7de2c1e7406717bb14332
>   Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
>   Date&#8203;:   Fri Jun 7 20&#8203;:16&#8203;:23 2013 \-0700
> 
>       \[perl \#24482\] Fix sort and require to treat CORE&#8203;:: as keyword
> 
> Looks like the code assumes that CORE&#8203;::\.\.\. is never bogus?

It adjusts the length by subtracting 6 \(length\("CORE&#8203;::"\)\) but does not set it back if it does not find a keyword\.

$ \./perl \-Ilib \-le 'sub C \{ $b \<=> $a \} print for sort CORE&#8203;::f 1\.\.10'
10
9
8
7
6
5
4
3
2
1

How do we want this to behave?  Normally require and sort treat a word as a bareword \(and not a sub call\) unless it is a keyword\.  It’s just that before 01b5ef509f CORE&#8203;::lc was not treated as a keyword\.

If we have CORE&#8203;::elsewhat should we revert to the previous behaviour of treating it as a bareword \(a sub name\)?  Or should we explicitly forbid it the way we do with CORE&#8203;::elsewhat elsewhere?

Either way it’s going to croak\, and the former is easier\.  \(Also\, I have my doubts as to whether croaking for CORE&#8203;::thing in general is a wise decision\, especially if core keywords are just special subroutines\.\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Dec 11 17&#8203;:55&#8203;:48 2014\, sprout wrote&#8203;:
> On Thu Dec 11 10&#8203;:45&#8203;:12 2014\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> > \* "Matthew Horsfall \(alh\)" \<wolfsage@&#8203;gmail\.com> \[2014\-12\-11T12&#8203;:00&#8203;:01\]
> > > Looks like this broke between 5\.19\.0 and 5\.19\.1&#8203;:
> >
> > Unsurprisingly\, it bisects to
> > 01b5ef509f2ebf466fd7de2c1e7406717bb14332
> >
> > commit 01b5ef509f2ebf466fd7de2c1e7406717bb14332
> > Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
> > Date&#8203;:   Fri Jun 7 20&#8203;:16&#8203;:23 2013 \-0700
> >
> > \[perl \#24482\] Fix sort and require to treat CORE&#8203;:: as keyword
> >
> > Looks like the code assumes that CORE&#8203;::\.\.\. is never bogus?
> 
> It adjusts the length by subtracting 6 \(length\("CORE&#8203;::"\)\) but does not
> set it back if it does not find a keyword\.
> 
> $ \./perl \-Ilib \-le 'sub C \{ $b \<=> $a \} print for sort CORE&#8203;::f 1\.\.10'
> 10
> 9
> 8
> 7
> 6
> 5
> 4
> 3
> 2
> 1
> 
> How do we want this to behave?  Normally require and sort treat a word
> as a bareword \(and not a sub call\) unless it is a keyword\.  It’s just
> that before 01b5ef509f CORE&#8203;::lc was not treated as a keyword\.
> 
> If we have CORE&#8203;::elsewhat should we revert to the previous behaviour
> of treating it as a bareword \(a sub name\)?  Or should we explicitly
> forbid it the way we do with CORE&#8203;::elsewhat elsewhere?
> 
> Either way it’s going to croak\, and the former is easier\.  \(Also\, I
> have my doubts as to whether croaking for CORE&#8203;::thing in general is a
> wise decision\, especially if core keywords are just special
> subroutines\.\)

I went with the former in 487e470dbd\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @rjbs

\* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2014\-12\-12T01&#8203;:29&#8203;:13\]
> I went with the former in 487e470dbd\.

\+1

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks for submitting this ticket

The issue should be resolved with the release today of Perl v5\.22\, available at http&#8203;://www\.perl\.org/get\.html
If you find that the problem persists\, feel free to reopen this ticket

\-\- 
Karl Williamson for the Perl 5 porters team

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
